### PR TITLE
Sync objects

### PR DESCRIPTION
**Pull request by astrozzc**
## Link(s) to Jira
- https://issues.redhat.com/browse/RHCLOUD-11197

## Description of Intent of Change(s)
Once the tenant_id relation is setup, we should sync the public and tenant schemas moving forward by ensure any non-get operation will affect both the public schema and the tenant schema.

This will keep feature and data parity moving forward, and should not affect the current process of creating tenants and storing/serving data from those tenant schemas. This will add an additional step of performing the same operation on the public schema.

## Local Testing
Unit tests

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [ ] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  -  if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: rbac/management/utils.py [hunk 1]
```diff
@@ -213,3 +216,19 @@ def roles_for_cross_account_principal(principal):
         )
         role_names_list = list(role_names)
     return Role.objects.filter(name__in=role_names_list)
+
+
+def create_object_in_tenant(schema_name, associate_tenant, model, **kwargs):
+    """Create object based on model in tenant schema."""
+    tenant = Tenant.objects.get(schema_name=schema_name)
+    with tenant_context(tenant):
+        model_object, created = model.objects.get_or_create(**kwargs)
```
Comment (coderbydesign) [inline rbac/management/utils.py:225]: After the reconciliation script is run, we'll want this to include the tenant in the get_or_create.

File: rbac/management/role/serializer.py [hunk 2]
```diff
@@ -337,3 +327,29 @@ def obtain_groups_in(obj, request):
...
 
     return Group.objects.filter(policies__in=policy_ids).distinct()
+
+
+def create_access_for_role(role, role_public, access_list, tenant):
+    """Create access objects and relate it to role."""
+    for access_item in access_list:
+        resource_def_list = access_item.pop("resourceDefinitions")
+        access_permission = access_item.pop("permission")
+        permission, _ = create_object_in_tenant(tenant.schema_name, tenant, Permission, **access_permission)
```
Comment (coderbydesign) [inline rbac/management/role/serializer.py:306]: Per @astrozzc, we'll want to make sure this is a get only.

**Comment by coderbydesign** [comment]
@astrozzc I don't think this works out of the box, but it would be awesome if we were able to use a similar pattern to this:
```python
import contextlib
from management.models import Principal
from api.models import Tenant

# this would provide as a handler to yield/run the same command
# on both the public schema, and the schema that's passed in,
# without changing the implementation logic in the original method
@contextlib.contextmanager
def schema_handler(tenant_schema):
    """Handle events in both public and tenant schemas."""
    public_schema = Tenant.objects.get(schema_name=schema_name)
    for schema in [public_schema, tenant_schema]:
        yield

# as we need to sync in both schemas, this could facilitate that
# by allowing us reuse the same code for storing/updating/deleting
# by calling a handler with the tenant schema
def some_other_method():
...
...
with schema_handler(Tenant.objects.get(schema_name="acct10001")):
    Principal.objects.get_or_create(username="username", cross_account=True)

# once we serve from public, we could just remove the `with schema_handler()`
# call in all places, after we set the schema to public, and not have to change
# any of the actual logic or calls
def some_other_method():
...
...
Principal.objects.get_or_create(username="username", cross_account=True)
```
Again, not sure how easy it would be, but something where we could basically just use a wrapper.

**Comment by astrozzc** [comment]
> @astrozzc I don't think this works out of the box, but it would be awesome if we were able to use a similar pattern to this:
> 
> ```python
> import contextlib
> from management.models import Principal
> from api.models import Tenant
> 
> # this would provide as a handler to yield/run the same command
> # on both the public schema, and the schema that's passed in,
> # without changing the implementation logic in the original method
> @contextlib.contextmanager
> def schema_handler(tenant_schema):
>     """Handle events in both public and tenant schemas."""
>     public_schema = Tenant.objects.get(schema_name=schema_name)
>     for schema in [public_schema, tenant_schema]:
>         yield
> 
> # as we need to sync in both schemas, this could facilitate that
> # by allowing us reuse the same code for storing/updating/deleting
> # by calling a handler with the tenant schema
> def some_other_method():
> ...
> ...
> with schema_handler(Tenant.objects.get(schema_name="acct10001")):
>     Principal.objects.get_or_create(username="username", cross_account=True)
> 
> # once we serve from public, we could just remove the `with schema_handler()`
> # call in all places, after we set the schema to public, and not have to change
> # any of the actual logic or calls
> def some_other_method():
> ...
> ...
> Principal.objects.get_or_create(username="username", cross_account=True)
> ```
> 
> Again, not sure how easy it would be, but something where we could basically just use a wrapper.

Yea, let me check. I think it would be cleaner to have such a method

**Review by coderbydesign — APPROVED** [review]
Nice work! This will make things much nicer to manage once we flip to the public schema. :) Functionally tested against:

roles:
- create
- update
- delete

groups:
- create
- update
- delete

- add/remove roles to/from groups
- add/remove principals to/from groups

**Pull request merged**